#15314: unify Booster speed across all ride types and increase limits

spacek531 opened the pull request:
This PR is structured with the idea that it will be merged before NTDF and be unlocked when it goes live, similar to many features unlocked by NSF.

spacek531 commented:
I am rethinking this PR, I think it should be implemented after https://github.com/OpenRCT2/OpenRCT2/issues/15417

spacek531 commented:
Scraping all of NEDesigns parks for the past 5 years reveal around 24 RCT2/OpenRCT2 parks that would be negatively affected by a simple change.

spacek531 commented:
Closing this because it is untenable to integrate parks that use hacks into the new schema.

spacek531 commented:
I've thought of a potential way to rescue this feature. On import of older parks, a VEHICLE_UPDATE_FLAG is set that recreates the legacy behavior. As it is currently, `get_booster_speed` takes the current ride type and speed of the booster and shifts it the appropriate amount:

```cpp
// Vehicle.cpp line 8081
auto boosterSpeed = get_booster_speed(curRide->type, (brake_speed << 16));
```
```cpp
// Ride.cpp line 5509
int32_t get_booster_speed(ride_type_t rideType, int32_t rawSpeed)
{
    int8_t shiftFactor = GetRideTypeDescriptor(rideType).OperatingSettings.BoosterSpeedFactor;
    if (shiftFactor == 0)
    {
        return rawSpeed;
    }
    if (shiftFactor > 0)
    {
        return (rawSpeed << shiftFactor);
    }

    // Workaround for an issue with older compilers (GCC 6, Clang 4) which would fail the build
    int8_t shiftFactorAbs = std::abs(shiftFactor);
    return (rawSpeed >> shiftFactorAbs);
}
```

I believe the burden of getting the booster speed should be shifted from UpdateTrackMotionForwards/Backwards to UpdateTrackMotionForwards/BackwardsGetNewTrack where it will interpret the track's absolute speed as absolute speed or relative speed based on the presence of the update flag.

```cpp
// Vehicle.cpp line 7955
brake_speed = GetBrakeBoosterSpeed(tileElement->AsTrack());
```
```cpp
// RideData.cpp
int32_t RideTypeDescriptor::GetRelativeSpeed(int32_t absoluteSpeed)
{
    if (OperatingSettings.BoosterSpeedFactor == 0)
    {
        return absoluteSpeed;
    }
    if (OperatingSettings.BoosterSpeedFactor > 0)
    {
        return (absoluteSpeed >> shiftFactor);
    }
    // Workaround for an issue with older compilers (GCC 6, Clang 4) which would fail the build
    int8_t shiftFactorAbs = std::abs(shiftFactor);
    return (rawSpeed << shiftFactorAbs);
}
int32_t RideTypeDescriptor::GetAbsoluteSpeed(int32_t relativeSpeed)
{
    if (OperatingSettings.BoosterSpeedFactor == 0)
    {
        return absoluteSpeed;
    }
    if (OperatingSettings.BoosterSpeedFactor > 0)
    {
        return (absoluteSpeed << shiftFactor);
    }
    // Workaround for an issue with older compilers (GCC 6, Clang 4) which would fail the build
    int8_t shiftFactorAbs = std::abs(shiftFactor);
    return (rawSpeed >> shiftFactorAbs);
}
```
```cpp
// Vehicle.cpp
int32_t Vehicle::GetBrakeBoosterSpeed(TrackElement *trackElement)
{
    auto speed= trackElement->GetBrakeBoosterSpeed();
    if (HasUpdateFlag(VEHICLE_UPDATE_FLAG_LEGACY_BOOSTER_SPEED)
    {
        speed = GetRide()->GetRideTypeDescriptor().GetAbsoluteSpeed(get_ride(trackElement->RideIndex)->GetRideTypeDescriptor().GetRelativeSpeed(speed));
    }
    return speed
}
```

spacek531 commented:
I believe I've managed to fully implement the feature outlined above, plus I extended it by getting the booster acceleration from the track's RTD. Getting booster acceleration from the track's RTD is temporary behavior which will be replaced by storing the BoosterAcceleration in the track element itself.

spacek531 commented on src/openrct2-ui/windows/RideConstruction.cpp:
```diff
@@ -2726,7 +2758,7 @@ static void WindowRideConstructionUpdateDisabledPieces(ObjectEntryIndex rideType
     if (rtd.HasFlag(RIDE_TYPE_FLAG_HAS_TRACK))
     {
-        // Set all pieces as “disabled”. When looping over the ride entries,
+        // Set all pieces as �disabled�. When looping over the ride entries,
```
These unicode chars keep getting munged and I don't know why. It's annoying to fix them, can we change them to regular bidirectional quotes?

spacek531 commented on src/openrct2-ui/windows/RideConstruction.cpp:
```diff
@@ -220,7 +220,10 @@ class RideConstructionWindow final : public Window
         _currentTrackPrice = MONEY32_UNDEFINED;
-        _currentBrakeSpeed2 = 8;
+#ifndef UNIFIED_BRAKE_SPEED_ENABLED
+        _currentBrakeSpeed2Temp = 3;
```
2Temp is an index value that represents the speed in _allowedBrakeSpeedsTemp array. It is used to limit the valid speeds to what .td6 can support.

spacek531 commented on src/openrct2-ui/windows/RideConstruction.cpp:
```diff
@@ -1284,9 +1287,11 @@ class RideConstructionWindow final : public Window
+#ifdef UNIFIED_BRAKE_SPEED_ENABLED
+                    // Increase brake speed
+                    uint8_t* brakesSpeedPtr = &_currentBrakeSpeed;
+                    uint8_t maxBrakesSpeed = std::numeric_limits<uint8_t>::max();
```
As a pre-requisite task, verify all RTD OperatingSettings.MaxBrakesSpeed and use the value for maxBrakesSpeed

spacek531 commented on src/openrct2-ui/windows/RideConstruction.cpp:
```diff
@@ -3058,7 +3090,35 @@ void WindowRideConstructionUpdateActiveElementsImpl()
+            {
+                _currentBrakeSpeed = tileElement->AsTrack()->GetBrakeBoosterSpeed();
+#ifndef UNIFIED_BRAKE_SPEED_ENABLED
+                if (tileElement->AsTrack()->GetTrackType() == TrackElemType::Booster)
```
This section forces the speed value to align with legacy speed values. Necessary for td6

spacek531 commented on src/openrct2/rct1/S4Importer.cpp:
```diff
@@ -2733,7 +2734,6 @@ namespace RCT1
         dst->num_seats = src->num_seats;
         dst->speed = src->speed;
         dst->powered_acceleration = src->powered_acceleration;
-        dst->brake_speed = src->brake_speed;
```
S4 and S6 import call UYppdateTrackBrakeSpeed() which calls GetSpeedFromTrackElement() which always sets brake_speed

spacek531 commented on src/openrct2/ride/Vehicle.cpp:
```diff
@@ -7608,25 +7610,22 @@ bool Vehicle::UpdateTrackMotionForwards(CarEntry* carEntry, Ride* curRide, rct_r
-    if ((trackType == TrackElemType::Flat && curRide->type == RIDE_TYPE_REVERSE_FREEFALL_COASTER)
-        || (trackType == TrackElemType::PoweredLift))
-    {
-        acceleration = GetRideTypeDescriptor(curRide->type).OperatingSettings.PoweredLiftAcceleration << 16;
```
OperatingSettings.PoweredLiftAcceleration is loaded into BoosterAcceleration and brake_speed is set to 255 under these conditions - this block becomes identical to the booster case.

spacek531 commented on src/openrct2/ride/Vehicle.cpp:
```diff
@@ -9335,6 +9334,39 @@ void Vehicle::EnableCollisionsForTrain()
+        return;
+    }
+    BoosterAcceleration = 0;
+    brake_speed = 0;
```
I believe there's potential desyncs if BoosterAcceleration and brake_speed are not set when importing from legacy.

spacek531 commented:
This PR has become slightly larger than anticipated with the inclusion of booster power, powered lift speed & power, and reverse freefall speed & power. I believe because of the limit of 16 bits in the vehicle updates flag, it's pertinent to do as much as possible in a single go, lest there become multiple versions of vehicle behaviours as each feature is added separately. When there's time after 0.4.3. drops I would like to work with someone on how this PR can be properly reviewed.

spacek531 commented on src/openrct2/GameStateSnapshots.cpp:
```diff
@@ -437,6 +437,7 @@ struct GameStateSnapshots final : public IGameStateSnapshots
         COMPARE_FIELD(Vehicle, powered_acceleration);
         COMPARE_FIELD(Vehicle, CollisionDetectionTimer);
         COMPARE_FIELD(Vehicle, animation_frame);
+        COMPARE_FIELD(Vehicle, BoosterAcceleration);
```
Todo: put this below the isCrashedVehicle bool where it belongs

spacek531 commented on src/openrct2/park/ParkFile.cpp:
```diff
@@ -162,6 +162,10 @@ namespace OpenRCT2
             }
+            if (os.GetHeader().TargetVersion < 17)
+            {
+                UpdateTrackBrakeSpeed();
```
This should be moved to somewhere after RTD objects would be loaded, if it is not already there. Likewise with S6 and S4z

spacek531 commented:
I think a testing plan could look something like this:

1. Build a park with all affected cases
- Rides for each BoosterSpeedFactor
- Rides for each BoosterAccelerationFactor
- Rides for each PoweredLiftAcceleration
- Reverse Freefall ride
- Merging onto booster of different BoosterSpeedFactor
- Merging onto booster of different BoosterAcceleratonFactor
- Reverse Freefall merging onto different PoweredLiftAcceleration

2. Branching from develop, add a uint8_t field at the end of Vehicle struct which is set to 0 at all times
3. Record replays from this special version
4. Analyze this branch's desync from the new replay

If the only desyncs are Vehicle struct values, then legacy behavior is preserved. If desyncs include vehicle position, legacy behavior was not preserved.

spacek531 commented:
Would solve https://github.com/OpenRCT2/OpenRCT2/issues/6680

spacek531 commented:
Please keep this open

karst commented:
Please keep open

karst commented:
Keep open please

karst commented:
Keep open

karst commented:
Still wip

spacek531 commented:
Todos:

- [ ] add plugin API
- [ ] beg @Basssiiie to update RVE
- [ ] add field to RTD to cap brake speed
- [ ] add field to RTD to cap booster speed
- [ ] make the "disable limits" cheat affect the caps

spacek531 commented on src/openrct2/ride/TrackDesign.cpp:
```diff
@@ -227,7 +227,27 @@ ResultWithMessage TrackDesign::CreateTrackDesignTrack(TrackDesignState& tds, con
+            {
+                trackFlags += 1;
+            }
+            trackFlags = std::min<uint8_t>(trackFlags, 0b00011111);
```
0b00011110

spacek531 commented on src/openrct2/ride/Vehicle.cpp:
```diff
@@ -7394,6 +7397,41 @@ void Vehicle::Sub6DBF3E()
+static uint8_t GetLegacyBoosterSpeed(uint8_t rawSpeed, RideTypeDescriptor& vehicleRTD, RideTypeDescriptor& trackRTD)
+{
+    auto relativeSpeed = trackRTD.GetRelativeBoosterSpeed(rawSpeed);
+    relativeSpeed &= 0b00011110; // legacy speeds have only 16 values, offset by 1 bit
```
0b00011110 should not be a magic number

spacek531 commented on src/openrct2/ride/Vehicle.cpp:
```diff
@@ -7663,9 +7702,8 @@ bool Vehicle::UpdateTrackMotionForwards(const CarEntry* carEntry, const Ride& cu
-            auto brakeSpeed = ChooseBrakeSpeed();
-
-            if ((brakeSpeed << 16) < _vehicleVelocityF64E08)
+            auto brakeSpeed = brake_speed << BrakeSpeedShiftAmount;
```
ChooseBrakeSpeed()

Basssiiie commented:
> - [ ] add plugin API
> - [ ] beg @Basssiiie to update RVE

@spacek531 Haha looking forward to that. 😁

Do you already know what kind of changes these are? Just a toggle for old vs new flag or is there more to take in account?

spacek531 commented:
> > * [ ]  add plugin API
> > * [ ]  beg @Basssiiie to update RVE
> 
> @spacek531 Haha looking forward to that. 😁
> 
> Do you already know what kind of changes these are? Just a toggle for old vs new flag or is there more to take in account?

I am going to copy the reverse car API as much as possible. It will be a simple toggle on/off. It is recommended to set the entire train to one or the other and not have any mixed in a ride.

Basssiiie commented:
> I am going to copy the reverse car API as much as possible. It will be a simple toggle on/off. It is recommended to set the entire ride to one or the other and not have any mixed in a ride.

If it's solely on/off for the entire ride and not intended to be mixed, I'd rather have it as a property on the ride than on the vehicle, from the plugin API's perspective.

spacek531 commented:
> > I am going to copy the reverse car API as much as possible. It will be a simple toggle on/off. It is recommended to set the entire ride to one or the other and not have any mixed in a ride.
> 
> If it's solely on/off for the entire ride and not intended to be mixed, I'd rather have it as a property on the ride than on the vehicle, from the plugin API's perspective.

I suppose that is true. I will look into what can be done with the ride. I will still add the API I described to the vehicle.

LordMarcel commented:
I did a whole bunch of tests on various massive and very exciting coasters with various types of boosters that I've built and found no difference in behaviour and resulting stats between the current development version and this debug version with the new booster behaviour.

karst commented:
Boost cap will definitely be a necessary addition.

When putting the numbers directly next to eachother I cannot hit the previous 215kmh that I had available before as a number.
It will instead put it to 217kmh 
![image](https://github.com/OpenRCT2/OpenRCT2/assets/1516893/28473d6b-8d2d-4182-bc54-e45ba7b6e8e1)

Here's in MPH
![image](https://github.com/OpenRCT2/OpenRCT2/assets/1516893/471e84e8-192c-4044-ad92-428f4e8498c6)

I do not know if this actually causes issues, though.

spacek531 commented:
The display speed is calculated by multiplying by 9 then bitshifting right. The values are double so the bitshift is increased by 1, changing the display speed. The vehicles should perform exactly the same.

spacek531 commented:
I currently have reserved 2 bits at the top of the BrakeBoosterSpeed field for future development of booster modes by @X123M3-256. There are 2 bits in the TrackColour field, though. TrackColour is currently host to a few properties. Bits 0-5 are used only on ghost train for the RCT1 terrain doors. Bits 0-4 are also used on multidim for seat rotation. Bits 6-7 are used for colour scheme 1-4. If the booster functionality utilizes bits 4-5 then both the unified speed feature and booster mode features can benefit.

Unified speed benefits from the full 8 bits in a few potential ways:
1. maximum speed is 2x giga booster top speed and is signed (which can benefit boosters and potentially block brakes in the future)
2. maximum speed is twice the unified precision  (matching the vehicle powered car speed increments) and 2x giga booster top speed
3. maximum speed is twice the unified precision and is signed
4. maximum speed is 4x giga booster top speed

Currently there are 4 different speed increments in the game, with each increment twice that of the previous: vehicle powered car speed, junior coaster booster speed, brake and regular booster speed, giga booster speed. With option 2 and 3, all these users will use the same intervals - vehicle powered car speed.

Option 3 I believe is the best way forward, but its benefits over option 2 will only be fully realized if TrackColour bits 4-5 can be used for booster behavior.

karst commented:
passive braking would be like reverse freefall rc? While active braking is like any other braking? Is the latter then also with a speed dial for brake speed? Or am I getting this wrong. Also speaking of reverse freefall rc, perhaps once something like this is implemented it's also an idea to convert the LIM-powered launch operating mode (which should really be called LSM-powered launch) to be a operating mode that allows for non-complete circuits without any special additions to it (do we have that already? It's like boat hire mode, but while allowing things to crash)?

spacek531 commented:
> passive braking would be like reverse freefall rc? While active braking is like any other braking? Is the latter then also with a speed dial for brake speed? Or am I getting this wrong. Also speaking of reverse freefall rc, perhaps once something like this is implemented it's also an idea to convert the LIM-powered launch operating mode (which should really be called LSM-powered launch) to be a operating mode that allows for non-complete circuits without any special additions to it (do we have that already? It's like boat hire mode, but while allowing things to crash)?

Reverse freefall RC and regular friction brakes braking force is linearly proportional to the vehicle's speed. Regular friction brakes are 8x more powerful than reverse freefall RC, so the deceleration is smoother, similar to eddy current brakes. @X123M3-256 how are your eddy-current brakes implemented?

The proposed "active" braking applies a constant force to the train whether it's going forward or reverse. If you remember from physics class, v1 = v0+a*t, that plots the behavior of the vehicle when on actively-braked boosters. The train will come to a stop and immediately go forward again.
![image](https://github.com/OpenRCT2/OpenRCT2/assets/3454156/009548e3-e812-48c2-b019-76d257b75791)

spacek531 commented:
Hey @Basssiiie my buddy, I've got a PR open that adds a vehicle flag. Would you consider updating your Ride Vehicle Editor plugin to support the flag? The new parameter is a boolean in Vehicle called `usesLegacyBoosterSpeed`. Many thanks!

spacek531 commented:
I have completed all of the features for the minimum viable product for unified brake and booster speeds. Before I call it ready for merge, I would like a decision on utilizing bits in TrackElement for storing booster mode. OpenRCT2's ride design capabilities are currently limited in part by the implementation of boosters - they are not keeping up with current roller coaster design trends of having boosters capable of forward and backwards launches. I also want Unified Brake and Booster Speeds to "get it right the first time" and use the same speed units as vehicle powered car speed. If the same speed units are not used, it creates an intermediate speed unit which will have to be rectified in the future, creating unnecessary import code bloat.

Previously, I suggested using bits 5 and 6 in TrackColour to store booster mode. This is not a perfect solution because there is potential interference with Ghost Train RCT1 terrain doors. I believe a better solution is to use bits 5 and 6 in Flags2. These bits have track piece-specific behavior. Bit 5 is the station green light flag, whose only user is the end station piece, and bit 6 is the brake closed bit, whose only users are brakes and block brakes. There is no overlap with boosters. No import or export code is required.

Approving this will bring the game in line with current roller coaster design philosophy and create a future-proof unified speed unit.

Basssiiie commented:
> Hey @Basssiiie my buddy, I've got a PR open that adds a vehicle flag. Would you consider updating your Ride Vehicle Editor plugin to support the flag? The new parameter is a boolean in Vehicle called `usesLegacyBoosterSpeed`. Many thanks!

Hey buddy. I will try to take a look at it this weekend. 

To double check: what are the specific effects on a vehicle when toggling this feature? Does it now boost according to track piece when set to `false`? And legacy behaviour means it boosts according to ride type? Are those all effects or is there more to it?

spacek531 commented:
> > Hey @Basssiiie my buddy, I've got a PR open that adds a vehicle flag. Would you consider updating your Ride Vehicle Editor plugin to support the flag? The new parameter is a boolean in Vehicle called `usesLegacyBoosterSpeed`. Many thanks!
> 
> Hey buddy. I will try to take a look at it this weekend.
> 
> To double check: what are the specific effects on a vehicle when toggling this feature? Does it now boost according to track piece when set to `false`? And legacy behaviour means it boosts according to ride type? Are those all effects or is there more to it?

Background: all boosters' speeds are converted into unified units, so they have the same precision as junior coaster boosters (currently - the precision will double when all 8 bits are available).

When a vehicle has the flag disabled and passes onto a booster, it will boost the vehicle to the speed set on the booster.

When a vehicle with the flag enabled passes onto a booster, the booster's speed is converted into "relative" units using the ride type descriptor of the *track* then masked. The relative units range 0-15. Then, the relative speed is converted into unified speed units the ride type descriptor of the *vehicle*. If a track has a non-legacy value, the speed is rounded down where applicable. If the RTD of the track and the RTD of the vehicle have the same legacy booster multiplier, the effect cancels out - giga coaster will divide track speed by 2 then multiply by 2, and junior will multiply track speed by 2 then divide by 2.

This process only applies to boosters, not brakes. Brakes will always slow the vehicle to the speed set on the booster, even if it is a non-legacy interval.

spacek531 commented:
- [ ] Make RideConstruction.cpp use the legacy intervals when building a ride with the Legacy lifecycle flag enabled.

spacek531 commented on src/openrct2/park/ParkFile.h:
```diff
@@ -9,10 +9,10 @@ struct ObjectRepositoryItem;
 {
     // Current version that is saved.
-    constexpr uint32_t PARK_FILE_CURRENT_VERSION = 32;
+    constexpr uint32_t PARK_FILE_CURRENT_VERSION = 33;
```
Todo: rebase and bump this

spacek531 commented on src/openrct2/ride/Track.cpp:
```diff
@@ -910,12 +915,12 @@ void TrackElement::SetIsIndestructible(bool isIndestructible)
 void TrackElement::SetBrakeBoosterSpeed(uint8_t speed)
 {
-    URide.BrakeBoosterSpeed = (speed >> 1);
+    URide.BrakeBoosterSpeed = speed & kMaximumBrakeSpeed;
```
Change this to unset the mask bits then bitwise-or the masked speed

Basssiiie commented:
@spacek531 Sorry it took a bit longer, but I've added the booster mode to the RideVehicleEditor: [see branch](https://github.com/Basssiiie/OpenRCT2-RideVehicleEditor/tree/booster-mode) 
Plugin download with the booster mode [RideVehicleEditor-v2.1-booster-mode.zip](https://github.com/OpenRCT2/OpenRCT2/files/13467132/RideVehicleEditor-v2.1-booster-mode.zip)

Dropdown:
![image](https://github.com/OpenRCT2/OpenRCT2/assets/20048660/af36b872-9b75-436c-a216-e92b606fb3e0)

Tooltip:
![image](https://github.com/OpenRCT2/OpenRCT2/assets/20048660/4129298c-e641-4f1e-bd68-415fed225c7f)

Not 100% sure about the wording used, but please let me know if fits. I assume it makes sense to have it available on all vehicles? Or is only the value from the first vehicle of the train used?

I can't currently test it as the CI hasn't created any artifacts yet.

spacek531 commented:
I have rebased and artifacts are generated now @Basssiiie 

Regarding the tooltip: the vehicle's ride determines the power and speed multiplier of boosters in legacy mode. In legacy mode, the booster power does not depend on the track at all, but the speed does. The junior coaster booster speed multiplier is 1x on junior coaster track, 1/4x on giga coaster track, and 1/2x on all other tracks. Giga coaster is the opposite: giga is 1x, junior is 4x, and all others are 2x. The property is read individually per vehicle in a train.

spacek531 commented:
Still waiting for a decision on adding behavior to Flags2 bits 5 and 6 (green light flag and brake closed flag)

tupaschoal commented:
> Still waiting for a decision on adding behavior to Flags2 bits 5 and 6 (green light flag and brake closed flag)

Do you have an opinion @Gymnasiast ?

Gymnasiast commented:
I’ll try and look at this with @spacek531 at some point (via DMs or an audio call) to make sure I fully understand it.

spacek531 commented on src/openrct2-ui/windows/RideConstruction.cpp:
```diff
@@ -217,9 +217,9 @@ static Widget _rideConstructionWidgets[] = {
-            _currentSeatRotationAngle = 4;
+        _currentTrackPrice = kMoney64Undefined;
+        _currentBrakeSpeed = 8;
+        _currentSeatRotationAngle = 4;
```
fix this indentation

spacek531 commented on src/openrct2-ui/windows/RideConstruction.cpp:
```diff
@@ -3037,21 +3029,23 @@ static Widget _rideConstructionWidgets[] = {
             }
+            _currentSeatRotationAngle = tileElement->AsTrack()->GetSeatRotation();
         }
+    }
```
Fix indentation

spacek531 commented:
Todo list:

- [ ] add API to ride object for ride lifecycle flag RIDE_LIFECYCLE_LEGACY_BOOSTER_SPEED
- [ ] re-add string STR_RIDE_SET_VEHICLE_LEGACY_SPEED_FAIL
- [ ] fix indentation where noted
- [ ] re-add brake speed cap
- [ ] add booster speed cap

spacek531 commented on src/openrct2-ui/windows/Ride.cpp:
```diff
@@ -265,10 +266,12 @@ static Widget _vehicleWidgets[] = {
     MakeWidget        ({  7, 154}, {302, 43}, WindowWidgetType::Scroll,   WindowColour::Secondary, STR_EMPTY                                         ),
     MakeSpinnerWidgets({  7, 203}, {145, 12}, WindowWidgetType::Spinner,  WindowColour::Secondary, STR_RIDE_VEHICLE_COUNT, STR_MAX_VEHICLES_TIP      ),
     MakeSpinnerWidgets({164, 203}, {145, 12}, WindowWidgetType::Spinner,  WindowColour::Secondary, STR_1_CAR_PER_TRAIN,    STR_MAX_CARS_PER_TRAIN_TIP),
+    MakeWidget        ({  310, 137}, {302, 12}, WindowWidgetType::Checkbox, WindowColour::Secondary, STR_OPTION_USE_LEGACY_SPEED, STR_OPTION_USE_LEGACY_SPEED_TIP  ),
```
investigate if this last entry is a rebase error or intentional

spacek531 commented on src/openrct2/park/ParkFile.cpp:
```diff
@@ -2130,6 +2166,7 @@ namespace OpenRCT2
                 entity.SetFlag(VehicleFlags::Crashed);
             }
         }
+
```
remove whitespace

Gymnasiast commented on src/openrct2/ride/RideConstruction.h:
```diff
@@ -62,7 +62,7 @@ extern TrackPitch _previousTrackPitchEnd;
 extern CoordsXYZ _previousTrackPiece;
 
-extern uint8_t _currentBrakeSpeed2;
+extern int8_t _currentBrakeSpeed;
```
Why change this to a signed integer?

spacek531 replied:
It was a change related to https://github.com/OpenRCT2/OpenRCT2/pull/16692#issuecomment-1801094250 that leaked into the original feature branch.

Gymnasiast replied:
Ok, can you revert it? I want to keep the diff as small as possible.

spacek531 replied:
I can, yes, but not for the next several days since I am very busy. I am still waiting on a decision regarding the 2 bits in the flags field as outlined in that comment.

Gymnasiast replied:
Sorry, didn’t know you expected me to weigh in on it. I’ll get to it.

Gymnasiast commented on data/language/en-GB.txt:
```diff
@@ -3699,6 +3699,9 @@ STR_6624    :Tile Inspector: Sort elements
 STR_6625    :Invalid colour
 STR_6626    :Animation is backwards
 STR_6627    :Track speed too high!
+STR_6628    :Track's speed is un-savable in current track design format. Speeds too high are capped to maximum, speeds at invalid intervals are rounded up.
```
```suggestion
STR_6628    :Track’s speed is un-savable in current track design format. Speeds too high are capped to maximum, speeds at invalid intervals are rounded up.
```

Gymnasiast commented on src/openrct2/ride/RideData.cpp:
```diff
@@ -454,3 +454,35 @@ TrackDrawerEntry getTrackDrawerEntry(const RideTypeDescriptor& rtd, bool isInver
+    {
+        return (absoluteSpeed >> shiftFactor);
+    }
+    // Workaround for an issue with older compilers (GCC 6, Clang 4) which would fail the build
```
We support neither of these any more. You can remove this comment (also from the other function). Optionally, you can also eliminate the intermediate variable `shiftFactorAbs`, but I’ll leave it to you to decide if you want that.

spacek531 commented on src/openrct2/localisation/StringIds.h:
```diff
@@ -1696,6 +1696,11 @@ enum : StringId
 
     STR_AT_LEAST_ONE_PEEP_NAMES_OBJECT_MUST_BE_SELECTED = 6676,
 
+    STR_TRACK_DESIGN_SPEED_UNSERIALISABLE = 6710,
```
TODO: move these to the ride strings section

spacek531 commented on src/openrct2/rct1/S4Importer.cpp:
```diff
@@ -886,6 +886,7 @@ namespace OpenRCT2::RCT1
             {
                 dst->lifecycle_flags |= RIDE_LIFECYCLE_REVERSED_TRAINS;
             }
+            dst->SetLifecycleFlag(RIDE_LIFECYCLE_LEGACY_BOOSTER_SPEED, true);
```
TODO: why is this 2000 lines above the other vehicle stuff?

spacek531 replied:
This is the Ride entity getting its legacy flag set

spacek531 commented on src/openrct2/rct2/S6Importer.cpp:
```diff
@@ -992,6 +992,7 @@ namespace OpenRCT2::RCT2
             dst->lift_hill_speed = src->LiftHillSpeed;
             dst->guests_favourite = src->GuestsFavourite;
             dst->lifecycle_flags = src->LifecycleFlags;
+            dst->SetLifecycleFlag(RIDE_LIFECYCLE_LEGACY_BOOSTER_SPEED, true);
```
TODO: why is the other vehicle stuff on line 2017

spacek531 replied:
This is the Ride entity getting its legacy flag set

spacek531 commented on src/openrct2/localisation/StringIds.h:
```diff
@@ -1696,6 +1696,11 @@ enum : StringId
     STR_AT_LEAST_ONE_PEEP_NAMES_OBJECT_MUST_BE_SELECTED = 6676,
 
+    STR_TRACK_DESIGN_SPEED_UNSERIALISABLE = 6710,
+    STR_RIDE_SET_VEHICLE_LEGACY_SPEED_FAIL = 6711,
```
TODO: remove this?

spacek531 commented on src/openrct2/ride/Vehicle.cpp:
```diff
@@ -7483,7 +7535,8 @@ bool Vehicle::UpdateTrackMotionBackwards(const CarEntry* carEntry, const Ride& c
         auto trackType = GetTrackType();
-        if (trackType == TrackElemType::Flat && curRide.GetRideTypeDescriptor().HasFlag(RtdFlag::hasLsmBehaviourOnFlat))
+        // Reverse Freefall Coaster braking
+        if (HasFlag(VehicleFlags::OnPoweredLift) && trackType == TrackElemType::Flat)
```
TODO: find a better way to do this that lays the foundation for booster modes.

spacek531 commented on src/openrct2-ui/windows/RideConstruction.cpp:
```diff
@@ -55,8 +55,8 @@
 
-constexpr int8_t kDefaultSpeedIncrement = 2;
-constexpr int8_t kDefaultMinimumSpeed = 2;
+constexpr int8_t kDefaultSpeedIncrement = 1;
```
What are people's thoughts on forcing valid speeds on rides with the legacy flag? I think we shouldn't. Invalid speeds are rounded down to the next valid legacy speed which isn't a major issue, and people should be using the new system anyway.

spacek531 commented:
My new plan is to implement features piecemeal instead of having them shoved into a single pull request. This is facilitated by https://github.com/OpenRCT2/OpenRCT2/pull/25920 which adds the flag that determines if the ride's vehicles obey the new behavior or legacy behavior.

In this PR, I had added both a ride flag and a vehicle flag. I've changed my mind about the vehicle flag, since some communities love abusing every possible implementation detail. To keep things easier to manage in the future, a ride's vehicles will obey its ride's flag. This means that it isn't possible to create an orphan vehicle that uses legacy behavior, but them's the breaks.

I haven't decided exactly how this PR will be broken up into sub-PRs. I have decided that this PR will be the final PR, and when the booster feature is fully implemented, this PR will remove the "development" aspect of https://github.com/OpenRCT2/OpenRCT2/pull/25920 and rebrand the flag as the legacy behavior flag, and add the changelog entry which points to https://github.com/OpenRCT2/OpenRCT2/discussions/23119

spacek531 commented on src/openrct2/ride/Vehicle.cpp:
Why is BlockBrakeSpeed set here? Something to do with flat track on reverse freefall? ~~Also this expensive check should not be done twice.~~ <- thought it was calling a different function

EDIT: OnPoweredLift behavior doesn't use brake_speed and BlockBrakeSpeed. It's setting brake_speed and BlockBrakeSpeed because otherwise they are unset, which in my opinion is a bad practice. Leaving it unset means it will stay at whatever it was last, which may have undesirable side effects.

spacek531 commented on src/openrct2/ride/Vehicle.cpp:
```suggestion
constexpr uint8_t const kBrakeSpeedShiftAmount = 16;
```

spacek531 commented on src/openrct2/ride/Vehicle.cpp:
```suggestion
constexpr uint8_t const kBrakeSpeedShiftAmount = 16;
constexpr uint8_t const kBoosterAccelerationShiftAmount = 16;
```

Gymnasiast commented:
As discussed on the voice chat, I’m closing this PR in favour of the physics-update branch.
